src/game/memory.h

```diff
@@ -48,16 +40,113 @@ void *segmented_to_virtual(const void *addr);
 void *virtual_to_segmented(u32 segment, const void *addr);
 void move_segment_table_to_dmem(void);
 
-void main_pool_init(void *start, void *end);
-void *main_pool_alloc(u32 size, u32 side);
-u32 main_pool_free(void *addr);
-void *main_pool_realloc(void *addr, u32 size);
+struct MainPoolRegion {
+    u8* start;
+    u8* end;
+};
+
+#define MAIN_POOL_REGIONS_COUNT 2
+
+struct MainPoolContext {
+    struct MainPoolRegion regions[MAIN_POOL_REGIONS_COUNT];
+};
+
+extern struct MainPoolContext gMainPool;
+
+#define MAIN_POOL_ALIGNMENT_DISABLE -1
+#define MAIN_POOL_ALLOC_TRY TRUE
+#define MAIN_POOL_ALLOC_FORCE FALSE
+
+// takes the first 'size' bytes from 'region'
+void main_pool_alloc_failed(int region, u32 size, int freeable, int line, const char* file) __attribute__((noreturn));
+
+// main_pool_region_alloc_from_start is a main entry point for all possible allocs in main pool.
+#define main_pool_region_alloc_from_start(region_idx, size, alignment, try) _main_pool_region_alloc_from_start(region_idx, size, alignment, try, __LINE__, __FILE__)
+static ALWAYS_INLINE void* _main_pool_region_alloc_from_start(int region_idx, u32 size, s32 alignment, int try, int line, const char* file) {
+    struct MainPoolRegion* region = &gMainPool.regions[region_idx];
+    u8* ret = alignment < 0 ? region->start : (u8*) ALIGN(region->start, alignment);
+    u8* new_start = ret + size;
+
+    if (try) {
+        if (new_start > region->end)
+            return NULL;
+    } else {
+#ifdef DEBUG_MEMORY_ALLOC
+        if (new_start > region->end)
+            main_pool_alloc_failed(region_idx, size, FALSE, line, file);
+#else
+        (void) line; (void) file;
+#endif
+    }
+
+    region->start = new_start;
+    if (!ret) __builtin_unreachable();
+    return ret;
+}
+
+/*
+ Main Pool is a trivial allocator that is managing multiple 'regions' of memory.
+ 'Region' is a contiguous block of memory available for main pool use.
+ It behaves similarly to an array of AllocOnly pools from vanilla SM64 by
+ "cutting" the start of the "region" when an allocation is made and returning the
+ pointer to the start of the initial "region".
+
+ Here is a simple visual example of how the memory is laid out in the main pool:
+
+ Main pool initial state is a multiple regions of memory:
+ |-------|   |----|  |-------------------|
+ If alloc(sizeof(+++++)), first region is used and the state becomes:
+ |+++++--|   |----|  |-------------------|
+  ^
+  returned pointer, no extra memory overhead
```

Comment:
The segmented portion was removed, but this comment still poses more questions than answers. Can I still allocate to the other memory regions? Are the other regions even supported in the first place?

Reply:
Yes, it can if you explicitly call `main_pool_region_alloc_from_start`

Reply:
I would add an example to this then to demonstrate that use case.

Reply:
Added a sample for region allocations